### PR TITLE
v0.3.0 Release Candidate

### PR DESCRIPTION
**Pull request by allow-cookies**
This pull request includes complete code refactoring to increase cohesion and get rid of coupling between serializers and views. Also the business logic was scattered across serializers, views and models - now the business logic lives in the commands, so responsibility segregation was increased.

More detailed list of changes:

* Support for Python >=3.6 only (older versions are not maintained anymore).
* Abandoned support of the `django-rest-framework-jwt` library as it is no longer maintained.
* Used `black` tool for auto code formatting.
* Type hints added.
* Backwards-compatible code refactoring and cleanup for better maintenance experience.
* Deactivation of primary MFA method would raise an error instead of randomly selecting new primary method from all active ones.
* Twilio backend client should be configured via environment variables instead of backend's configuration the `TWILIO_ACCOUNT_SID` and `TWILIO_AUTH_TOKEN`.

### CODE REVIEW

File: testproject/settings.py [hunk 1]
```diff
@@ -82,23 +58,16 @@
...
 }
 
-
-# Password validation
-# https://docs.djangoproject.com/en/1.11/ref/settings/#auth-password-validators
-
 AUTH_PASSWORD_VALIDATORS = [
     {
-        "NAME": "django.contrib.auth.password_validation.UserAttributeSimilarityValidator",
+        "NAME": "django.contrib.auth.password_validation.UserAttributeSimilarityValidator",  # noqa: E501
```
Comment (lukzmu) [inline testproject/settings.py:70]: I really like the `E501` apart from `noqa` 👍🏻.

**Review by lukzmu — APPROVED** [review]
Approved. One thing you could do, is to fix the `codecov` drop, so the pipeline does not fail.

File: testproject/tests/test_first_step_authentication.py [hunk 2]
```diff
@@ -16,11 +16,19 @@ def test_get_emphemeral_token(active_user_with_email_otp):
     response = login(active_user_with_email_otp)
     assert response.status_code == 200
```
Comment (mwrobel-merixstudio) [inline testproject/tests/test_first_step_authentication.py:17]: in other tests we are using `rest_framework.status` const so we can do in this test too

File: testproject/tests/test_first_step_authentication.py [hunk 3]
```diff
@@ -16,11 +16,19 @@ def test_get_emphemeral_token(active_user_with_email_otp):
...
+            user=None, token=response.data.get("ephemeral_token")
+        )
         == active_user_with_email_otp
     )
 
 
+@pytest.mark.django_db
+def test_deactivated_user(deactivated_user_with_email_otp):
+    response = login(deactivated_user_with_email_otp)
+    assert response.status_code == 400
```
Comment (mwrobel-merixstudio) [inline testproject/tests/test_first_step_authentication.py:29]: in other tests we are using `rest_framework.status` const so we can do in this test too

File: testproject/tests/test_first_step_authentication.py [hunk 4]
```diff
@@ -89,16 +97,4 @@ def test_login_wrong_password(active_user):
         format="json",
     )
     assert response.status_code == 400
```
Comment (mwrobel-merixstudio) [inline testproject/tests/test_first_step_authentication.py:99]: in other tests we are using `rest_framework.status` const so we can do in this test too

File: testproject/tests/test_second_step_authentication.py [hunk 5]
```diff
@@ -3,31 +3,65 @@
...
+    response = client.post(
+        path=PATH_AUTH_JWT_LOGIN_CODE,
+        data={
+            "ephemeral_token": first_step.data.get("ephemeral_token"),
+            "code": handler.create_code(),
+        },
+        format="json",
+    )
+
+    assert response.status_code == 200
```
Comment (mwrobel-merixstudio) [inline testproject/tests/test_second_step_authentication.py:52]: in other tests we are using `rest_framework.status` const so we can do in this test too

File: testproject/tests/test_second_step_authentication.py [hunk 6]
```diff
@@ -3,31 +3,65 @@
...
+    time.sleep(3)
+    response = client.post(
+        path=PATH_AUTH_JWT_LOGIN_CODE,
+        data={
+            "ephemeral_token": first_step.data.get("ephemeral_token"),
+            "code": code,
+        },
+        format="json",
+    )
+    assert response.status_code == 401
```
Comment (mwrobel-merixstudio) [inline testproject/tests/test_second_step_authentication.py:42]: in other tests we are using `rest_framework.status` const so we can do in this test too

File: testproject/tests/test_second_step_authentication.py [hunk 7]
```diff
@@ -60,16 +94,15 @@ def test_wrong_second_step_verification_with_wrong_code(active_user_with_email_o
...
         data={
             "ephemeral_token": first_step.data.get("ephemeral_token"),
             "code": "test",
         },
         format="json",
     )
-    code_error = "invalid_code"
-    assert response.status_code == 400
-    assert response.data.get("non_field_errors")[0].code == code_error
+    assert response.status_code == 401
```
Comment (mwrobel-merixstudio) [inline testproject/tests/test_second_step_authentication.py:104]: in other tests we are using `rest_framework.status` const so we can do in this test too

File: testproject/tests/test_second_step_authentication.py [hunk 8]
```diff
@@ -78,28 +111,28 @@ def test_wrong_second_step_verification_with_ephemeral_token(
...
+        path=PATH_AUTH_JWT_LOGIN_CODE,
         data={
             "ephemeral_token": first_step.data.get("ephemeral_token") + "wrong",
-            "code": create_otp_code(secret),
+            "code": handler.create_code(),
         },
         format="json",
     )
-    assert response.status_code == 400
+    assert response.status_code == 401
```
Comment (mwrobel-merixstudio) [inline testproject/tests/test_second_step_authentication.py:123]: in other tests we are using `rest_framework.status` const so we can do in this test too

File: testproject/tests/test_second_step_authentication.py [hunk 9]
```diff
@@ -151,22 +182,21 @@ def test_second_method_activation_already_active(active_user_with_email_otp):
         path="/auth/email/activate/",
         format="json",
     )
-    error_msg = "MFA method already active."
     assert response.status_code == 400
```
Comment (mwrobel-merixstudio) [inline testproject/tests/test_second_step_authentication.py:185]: in other tests we are using `rest_framework.status` const so we can do in this test too

File: testproject/tests/test_second_step_authentication.py [hunk 10]
```diff
@@ -198,14 +228,15 @@ def test_activation_otp_confirm_wrong(active_user):
         path="/auth/email/activate/",
         format="json",
     )
+    assert response.status_code == 200
     response = client.post(
         path="/auth/email/activate/confirm/",
         data={"code": "test00"},
         format="json",
     )
     assert response.status_code == 400
```
Comment (mwrobel-merixstudio) [inline testproject/tests/test_second_step_authentication.py:237]: in other tests we are using `rest_framework.status` const so we can do in this test too

File: testproject/tests/test_second_step_authentication.py [hunk 11]
```diff
@@ -363,63 +355,28 @@ def test_new_method_after_deactivation_same_method_wrong(
...
-            get_token_from_response(login_response)
-        )
-    )
-    response = client.post(
-        path="/auth/email/deactivate/",
-        data={"new_primary_method": "test"},
-        format="json",
-    )
-
     assert response.status_code == 400
```
Comment (mwrobel-merixstudio) [inline testproject/tests/test_second_step_authentication.py:364]: in other tests we are using `rest_framework.status` const so we can do in this test too

File: testproject/tests/test_second_step_authentication.py [hunk 12]
```diff
@@ -431,17 +388,17 @@ def test_change_primary_method(active_user_with_many_otp_methods):
...
         },
         format="json",
     )
     new_primary_method = active_user.mfa_methods.filter(
         is_primary=True,
     )[0]
-    assert response.status_code == 200
-    assert first_primary_method != new_primary_method
-    assert new_primary_method.name == "sms"
+    assert response.status_code == 204
```
Comment (mwrobel-merixstudio) [inline testproject/tests/test_second_step_authentication.py:399]: in other tests we are using `rest_framework.status` const so we can do in this test too

File: testproject/tests/test_second_step_authentication.py [hunk 13]
```diff
@@ -470,32 +426,31 @@ def test_change_primary_method_with_backup_code(
...
+            "method": "sms_twilio",
             "code": backup_code,
         },
         format="json",
     )
     new_primary_method = active_user.mfa_methods.filter(
         is_primary=True,
     )[0]
-    assert response.status_code == 200
+    assert response.status_code == 204
```
Comment (mwrobel-merixstudio) [inline testproject/tests/test_second_step_authentication.py:437]: in other tests we are using `rest_framework.status` const so we can do in this test too

File: testproject/tests/test_second_step_authentication.py [hunk 14]
```diff
@@ -507,28 +462,28 @@ def test_change_primary_method_to_invalid_wrong(active_user_with_many_otp_method
...
         path="/auth/mfa/change-primary-method/",
         data={
-            "method": "sms",
+            "method": "sms_twilio",
             "code": "test",
         },
         format="json",
     )
-    error_code = "invalid_code"
     assert response.status_code == 400
```
Comment (mwrobel-merixstudio) [inline testproject/tests/test_second_step_authentication.py:470]: in other tests we are using `rest_framework.status` const so we can do in this test too

File: testproject/tests/test_second_step_authentication.py [hunk 15]
```diff
@@ -540,21 +495,20 @@ def test_change_primary_method_to_inactive(active_user_with_email_otp):
...
         data={
-            "method": "sms",
-            "code": create_otp_code(first_primary_method.secret),
+            "method": "sms_twilio",
+            "code": handler.create_code(),
         },
         format="json",
     )
-    error_code = "missing_method"
     assert response.status_code == 400
```
Comment (mwrobel-merixstudio) [inline testproject/tests/test_second_step_authentication.py:503]: in other tests we are using `rest_framework.status` const so we can do in this test too

File: testproject/tests/test_second_step_authentication.py [hunk 16]
```diff
@@ -572,29 +526,28 @@ def test_change_primary_disabled_method_wrong(active_user):
...
         path="/auth/mfa/change-primary-method/",
         data={
-            "method": "sms",
+            "method": "sms_twilio",
             "code": "code",
         },
         format="json",
     )
-    error_code = "not_enabled"
     assert response.status_code == 400
```
Comment (mwrobel-merixstudio) [inline testproject/tests/test_second_step_authentication.py:534]: in other tests we are using `rest_framework.status` const so we can do in this test too

File: testproject/tests/test_second_step_authentication.py [hunk 17]
```diff
@@ -652,26 +606,25 @@ def test_request_codes(active_user_with_email_otp):
...
     response = client.post(
         path="/auth/code/request/",
-        data={
-            "method": "email",
-        },
+        data={"method": "email"},
         format="json",
     )
     expected_msg = "Email message with MFA code has been sent."
     assert response.status_code == 200
```
Comment (mwrobel-merixstudio) [inline testproject/tests/test_second_step_authentication.py:613]: in other tests we are using `rest_framework.status` const so we can do in this test too

File: testproject/tests/test_second_step_authentication.py [hunk 18]
```diff
@@ -809,39 +763,87 @@ def test_backup_codes_regeneration_disabled_method(
...
     response = client.post(
-        path="/auth/sms/codes/regenerate/",
+        path="/auth/sms_twilio/codes/regenerate/",
+        data={
+            "code": handler.create_code(),
+        },
         format="json",
     )
-    error_msg = "Method is disabled."
     assert response.status_code == 400
```
Comment (mwrobel-merixstudio) [inline testproject/tests/test_second_step_authentication.py:772]: in other tests we are using `rest_framework.status` const so we can do in this test too

File: testproject/tests/test_second_step_authentication.py [hunk 19]
```diff
@@ -809,39 +763,87 @@ def test_backup_codes_regeneration_disabled_method(
...
+    handler = get_mfa_handler(mfa_method=active_user_with_yubi.mfa_methods.first())
+    response = APIClient().post(
+        path=PATH_AUTH_JWT_LOGIN_CODE,
+        data={
+            "ephemeral_token": first_step_response.data.get("ephemeral_token"),
+            "code": handler.create_code(),
+        },
         format="json",
     )
     assert response.status_code == 200
```
Comment (mwrobel-merixstudio) [inline testproject/tests/test_second_step_authentication.py:788]: in other tests we are using `rest_framework.status` const so we can do in this test too

File: testproject/tests/test_second_step_authentication.py [hunk 20]
```diff
@@ -809,39 +763,87 @@ def test_backup_codes_regeneration_disabled_method(
...
+    handler = get_mfa_handler(mfa_method=active_user_with_yubi.mfa_methods.first())
+    response = APIClient().post(
+        path=PATH_AUTH_JWT_LOGIN_CODE,
+        data={
+            "ephemeral_token": first_step_response.data.get("ephemeral_token"),
+            "code": handler.create_code(),
+        },
+        format="json",
+    )
+    assert response.status_code == 401
```
Comment (mwrobel-merixstudio) [inline testproject/tests/test_second_step_authentication.py:803]: and so on ;) so you can just use `rest_framework.status` consts in all tests on each `status_code` assertion in all tests

**Review by mwrobel-merixstudio — CHANGES_REQUESTED** [review]
everything looks fine, one last thing - we can drop this explicit assertions on `status_code` and use `rest_framework.status` consts

**Review by mwrobel-merixstudio — APPROVED** [review]
looks good :)

**Pull request merged**